<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="heinin" />

<meta name="date" content="2024-02-05" />

<title>Aggregating ASE across multiple SNPs</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ILD_ASE_Xenium</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/heinin/ILD_ASE_Xenium">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Aggregating ASE across multiple SNPs</h1>
<h4 class="author">heinin</h4>
<h4 class="date">2024-02-05</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-03-14
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>ILD_ASE_Xenium/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20231206code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20231206)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20231206code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20231206)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it
difficult for you and others to run your code on a different machine.
Change the absolute path(s) below to the suggested relative path(s) to
make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
/home/hnatri/ILD_ASE_Xenium/
</td>
<td style="text-align:left;">
.
</td>
</tr>
<tr>
<td style="text-align:left;">
/home/hnatri/ILD_ASE_Xenium/code/colors_themes.R
</td>
<td style="text-align:left;">
code/colors_themes.R
</td>
</tr>
<tr>
<td style="text-align:left;">
/home/hnatri/ILD_ASE_Xenium/code/plot_functions.R
</td>
<td style="text-align:left;">
code/plot_functions.R
</td>
</tr>
<tr>
<td style="text-align:left;">
/home/hnatri/ILD_ASE_Xenium/code/utilities.R
</td>
<td style="text-align:left;">
code/utilities.R
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsheininghpLRgVwS7Z48swjRrQoBOAaVz2kQoFlP0cEwwugithubcomheininILDASEXeniumtreeb33c6bf8af2033222b12f4cd4511c6d4c7be68fbtargetblankb33c6bfa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://heinin: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@github.com/heinin/ILD_ASE_Xenium/tree/b33c6bf8af2033222b12f4cd4511c6d4c7be68fb" target="_blank">b33c6bf</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsheininghpLRgVwS7Z48swjRrQoBOAaVz2kQoFlP0cEwwugithubcomheininILDASEXeniumtreeb33c6bf8af2033222b12f4cd4511c6d4c7be68fbtargetblankb33c6bfa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://heinin: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@github.com/heinin/ILD_ASE_Xenium/tree/b33c6bf8af2033222b12f4cd4511c6d4c7be68fb" target="_blank">b33c6bf</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/aggregate_multiSNP.Rmd

Unstaged changes:
    Modified:   analysis/ASE.Rmd
    Modified:   analysis/Xenium_preprocessing.Rmd
    Modified:   analysis/ase_readcounts.Rmd
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="packages-and-environment-variables" class="section level3">
<h3>Packages and environment variables</h3>
<pre class="r"><code>suppressPackageStartupMessages({library(cli)
                                library(Seurat)
                                library(SeuratObject)
                                library(SeuratDisk)
                                library(tidyverse)
                                library(tibble)
                                library(plyr)
                                library(dplyr)
                                library(ggplot2)
                                library(ggpubr)
                                library(ggrepel)
                                library(workflowr)
                                library(googlesheets4)
                                library(VariantAnnotation)})

setwd(&quot;/home/hnatri/ILD_ASE_Xenium/&quot;)
set.seed(9999)
options(ggrepel.max.overlaps = Inf)

# Colors, themes, cell type markers, and plot functions
source(&quot;/home/hnatri/ILD_ASE_Xenium/code/colors_themes.R&quot;)
source(&quot;/home/hnatri/ILD_ASE_Xenium/code/plot_functions.R&quot;)
source(&quot;/home/hnatri/ILD_ASE_Xenium/code/utilities.R&quot;)

# Cell type annotations
gs4_deauth()
ct_annot  &lt;- gs4_get(&quot;https://docs.google.com/spreadsheets/d/1SDfhxf6SjllxXEtNPf32ZKTEqHC9QJW3BpRYZFhpqFE/edit?usp=sharing&quot;)
sheet_names(ct_annot)
all_celltypes_annot &lt;- read_sheet(ct_annot, sheet = &quot;All celltypes, annotated, merged&quot;)</code></pre>
</div>
<div id="importing-data" class="section level3">
<h3>Importing data</h3>
<pre class="r"><code>seurat_object &lt;- readRDS(&quot;/tgen_labs/banovich/IPF/Spatial_ASE/ILD_ASE_Xenium_annotated.rds&quot;)</code></pre>
</div>
<div id="importing-eqtl-and-tag-snp-info" class="section level3">
<h3>Importing eQTL and tag-SNP info</h3>
<pre class="r"><code>gs4_deauth()
gtables  &lt;- gs4_get(&quot;https://docs.google.com/spreadsheets/d/1f7HtQC07PWa1ij9UxTmgrEMX2OnHDVcrpn7Oa_VNIms/edit?usp=sharing&quot;)
sheet_names(gtables)</code></pre>
<pre><code>[1] &quot;More candidates, SNP info with exon number&quot;      
[2] &quot;Genotypes for selected loci&quot;                     
[3] &quot;Genotypes for selected loci, FFPE blocks at TGen&quot;
[4] &quot;SNP info&quot;                                        
[5] &quot;Updated candidates&quot;                              
[6] &quot;Candidates&quot;                                      
[7] &quot;Sheet7&quot;                                          
[8] &quot;Copy of Sheet7&quot;                                  </code></pre>
<pre class="r"><code>ld_snp_info &lt;- read_sheet(gtables, sheet = &quot;More candidates, SNP info with exon number&quot;)</code></pre>
<pre><code>✔ Reading from &quot;ILD spatial ASE&quot;.</code></pre>
<pre><code>✔ Range &#39;&#39;More candidates, SNP info with exon number&#39;&#39;.</code></pre>
<pre class="r"><code>ld_snp_info &lt;- as.data.frame(ld_snp_info)

# Adding probe names
ld_snp_info$probe &lt;- paste(ld_snp_info$gene, ld_snp_info$chr, ld_snp_info$linked_snp_pos, ld_snp_info$alt, sep = &quot;-&quot;)
ld_snp_info$probe1 &lt;- paste(ld_snp_info$gene, ld_snp_info$chr, ld_snp_info$linked_snp_pos, ld_snp_info$ref, sep = &quot;-&quot;)
ld_snp_info$probe2 &lt;- paste(ld_snp_info$gene, ld_snp_info$chr, ld_snp_info$linked_snp_pos, ld_snp_info$alt, sep = &quot;-&quot;)</code></pre>
</div>
<div id="selecting-the-genes-to-plot" class="section level3">
<h3>Selecting the genes to plot</h3>
<pre class="r"><code># Obtaining the expression matrix
pseudobulk &lt;- AverageExpression(seurat_object,
                                group.by = &quot;Sample_Name&quot;,
                                assay = &quot;RNA&quot;,
                                slot = &quot;counts&quot;)</code></pre>
<pre><code>Warning: The `slot` argument of `AverageExpression()` is deprecated as of Seurat 5.0.0.
ℹ Please use the `layer` argument instead.
ℹ The deprecated feature was likely used in the Seurat package.
  Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<pre><code>As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.
This message is displayed once per session.</code></pre>
<pre class="r"><code># Calculating counts per million
#pseudobulk &lt;- RelativeCounts(pseudobulk$RNA, scale.factor = 1e6)

pseudobulk &lt;- as.data.frame(t(pseudobulk$RNA))

ld_snp_info &lt;- ld_snp_info %&gt;% filter(probe1 %in% colnames(pseudobulk) |
                                        probe2 %in% colnames(pseudobulk))</code></pre>
</div>
<div id="aggregating-all-refalt-counts" class="section level3">
<h3>Aggregating all REF/ALT counts</h3>
<pre class="r"><code># Assembling a dataframe with a gene-tag-SNP pair, gene, probe1, and, probe2 
# columns with pseudobulk counts for each sample

# A list of plots for each gene-tag-SNP pair
ld_snp_info$gene_tagSNP &lt;- paste0(ld_snp_info$gene, &quot;_&quot;, ld_snp_info$gene_body_snp)
ld_snp_info &lt;- ld_snp_info %&gt;%
  dplyr::select(-c(&quot;R2&quot;, &quot;gene_snp&quot;, &quot;esnp_loc&quot;)) %&gt;% distinct()

data_list &lt;- lapply(ld_snp_info$gene_tagSNP, function(xx){
  gene &lt;- strsplit(xx, &quot;_&quot;)[[1]][1]
  tagSNP &lt;- strsplit(xx, &quot;_&quot;)[[1]][2]
  probe1 &lt;- ld_snp_info %&gt;% filter(gene_tagSNP == xx) %&gt;% dplyr::select(probe1) %&gt;% unique() %&gt;% as.character()
  probe2 &lt;- ld_snp_info %&gt;% filter(gene_tagSNP == xx) %&gt;% dplyr::select(probe2) %&gt;% unique() %&gt;% as.character()
  eSNP &lt;- ld_snp_info %&gt;% filter(gene_tagSNP == xx) %&gt;% dplyr::select(esnp) %&gt;% unique() %&gt;% as.character()
  exp_data &lt;- pseudobulk %&gt;% dplyr::select(c(probe1, probe2))
  #exp_data$probe_sum &lt;- exp_data[,probe1] + exp_data[,probe2]
  colnames(exp_data) &lt;- gsub(&quot;-&quot;, &quot;_&quot;, colnames(exp_data))
  probe1 &lt;- gsub(&quot;-&quot;, &quot;_&quot;, probe1)
  probe2 &lt;- gsub(&quot;-&quot;, &quot;_&quot;, probe2)
  colnames(exp_data) &lt;- gsub(probe1, &quot;REFprobeCounts&quot;, colnames(exp_data))
  colnames(exp_data) &lt;- gsub(probe2, &quot;ALTprobeCounts&quot;, colnames(exp_data))
  #colnames(exp_data) &lt;- c(&quot;REFprobeCounts&quot;, &quot;ALTprobeCounts&quot;)
  exp_data$gene &lt;- gene
  exp_data$tagSNP &lt;- tagSNP
  exp_data$eSNP &lt;- eSNP
  exp_data$Sample_Name &lt;- rownames(exp_data)
  rownames(exp_data) &lt;- NULL
 
  exp_data
})</code></pre>
<pre><code>Warning: Using an external vector in selections was deprecated in tidyselect 1.1.0.
ℹ Please use `all_of()` or `any_of()` instead.
  # Was:
  data %&gt;% select(probe1)

  # Now:
  data %&gt;% select(all_of(probe1))

See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<pre><code>Warning: Using an external vector in selections was deprecated in tidyselect 1.1.0.
ℹ Please use `all_of()` or `any_of()` instead.
  # Was:
  data %&gt;% select(probe2)

  # Now:
  data %&gt;% select(all_of(probe2))

See &lt;https://tidyselect.r-lib.org/reference/faq-external-vector.html&gt;.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<pre class="r"><code># To a dataframe
exp_data &lt;- do.call(&quot;rbind&quot;, data_list)</code></pre>
</div>
<div id="plotting.-each-plot-item-is-a-sampleprobe-pair."
class="section level3">
<h3>Plotting. Each plot item is a sample+probe pair.</h3>
<pre class="r"><code># Read counts stats for REF and ALT probes
summary(exp_data$REFprobeCounts)</code></pre>
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
 0.000000  0.002994  0.006795  0.358737  0.022135 13.588421 </code></pre>
<pre class="r"><code>summary(exp_data$ALTprobeCounts)</code></pre>
<pre><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
0.000000 0.002976 0.006564 0.293812 0.025663 7.242985 </code></pre>
<pre class="r"><code># Total numbers of reads mapping to REF and ALT probes
sum(exp_data$REFprobeCounts)</code></pre>
<pre><code>[1] 1349.569</code></pre>
<pre class="r"><code>sum(exp_data$ALTprobeCounts)</code></pre>
<pre><code>[1] 1105.32</code></pre>
<pre class="r"><code># Numbers of instances with average read count &gt;1
exp_data %&gt;% summarise(alt_below_one = count(ALTprobeCounts&lt;1),
                       ref_below_one = count(REFprobeCounts&lt;1),
                       alt_above_one = count(ALTprobeCounts&gt;1),
                       ref_above_one = count(REFprobeCounts&gt;1))</code></pre>
<pre><code>  alt_below_one ref_below_one alt_above_one ref_above_one
1          3386          3305           376           457</code></pre>
<pre class="r"><code># Scatterplot of read counts
exp_data %&gt;%
  ggplot(aes(x = REFprobeCounts, y = ALTprobeCounts)) + 
    geom_point() +
    theme_classic()</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Histogram of average read counts
exp_data %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                          names_to = &quot;probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  ggplot(aes(x = aveCounts, fill = probe)) + 
    geom_histogram(binwidth = 1) +
    theme_classic()</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># filtering out values &lt;1
exp_data %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                          names_to = &quot;probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  filter(aveCounts&gt;1) %&gt;%
  ggplot(aes(x = aveCounts, fill = probe)) + 
    geom_histogram(binwidth = 1) +
    theme_classic()</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_data %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                          names_to = &quot;probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  ggplot(aes(x = probe, y = log2(aveCounts), fill = probe)) +
    geom_violin() +
    theme_classic()</code></pre>
<pre><code>Warning: Removed 49 rows containing non-finite values (`stat_ydensity()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-4.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_data %&gt;% mutate(ratio_alt_ref = ALTprobeCounts/REFprobeCounts) %&gt;%
    ggplot(aes(x = ratio_alt_ref)) + 
      geom_histogram() +
      theme_classic()</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<pre><code>Warning: Removed 28 rows containing non-finite values (`stat_bin()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-5.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_data %&gt;% mutate(ratio_alt_ref = REFprobeCounts/ALTprobeCounts) %&gt;%
    ggplot(aes(x = ratio_alt_ref)) + 
      geom_histogram() +
      theme_classic()</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<pre><code>Warning: Removed 21 rows containing non-finite values (`stat_bin()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-6.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p1 &lt;- exp_data %&gt;% mutate(ratio_alt_ref = ALTprobeCounts/REFprobeCounts) %&gt;%
    ggplot(aes(x = ratio_alt_ref)) + 
      geom_histogram() +
      theme_classic() +
      xlim(c(0, 40)) +
      ylim(c(0, 1200))

p2 &lt;- exp_data %&gt;% mutate(ratio_ref_alt = REFprobeCounts/ALTprobeCounts) %&gt;%
    ggplot(aes(x = ratio_ref_alt)) + 
      geom_histogram() +
      theme_classic() +
      xlim(c(0, 40)) +
      ylim(c(0, 1200))

p1 + p2</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<pre><code>Warning: Removed 92 rows containing non-finite values (`stat_bin()`).</code></pre>
<pre><code>Warning: Removed 3 rows containing missing values (`geom_bar()`).</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<pre><code>Warning: Removed 25 rows containing non-finite values (`stat_bin()`).
Removed 3 rows containing missing values (`geom_bar()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-6-7.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="adding-genotype-information" class="section level3">
<h3>Adding genotype information</h3>
<pre class="r"><code># Genotype data from ASE.rds
gt_data &lt;- read.csv(&quot;/scratch/hnatri/ILD/ILD_spatial_ASE/gt_data.csv&quot;, row.names = &quot;X&quot;)

#gt_data$eSNP_GT &lt;- as.numeric(gt_data$eSNP_GT)
#gt_data$linked_SNP_GT &lt;- as.numeric(gt_data$linked_SNP_GT)

colnames(gt_data) &lt;- gsub(&quot;gene_body_snp&quot;, &quot;tagSNP&quot;, colnames(gt_data))

intersect(gt_data$Sample_Name, exp_data$Sample_Name)</code></pre>
<pre><code> [1] &quot;THD0016&quot;  &quot;THD0026&quot;  &quot;TILD001&quot;  &quot;TILD006&quot;  &quot;TILD010&quot;  &quot;TILD028&quot; 
 [7] &quot;TILD030&quot;  &quot;TILD037&quot;  &quot;TILD041&quot;  &quot;TILD049&quot;  &quot;TILD055&quot;  &quot;TILD059&quot; 
[13] &quot;TILD062&quot;  &quot;TILD074&quot;  &quot;TILD080&quot;  &quot;TILD093&quot;  &quot;TILD102&quot;  &quot;TILD103&quot; 
[19] &quot;TILD109&quot;  &quot;TILD111&quot;  &quot;TILD113&quot;  &quot;TILD123&quot;  &quot;TILD126&quot;  &quot;TILD136&quot; 
[25] &quot;VUHD113&quot;  &quot;VUHD115&quot;  &quot;VUILD102&quot; &quot;VUILD104&quot; &quot;VUILD48&quot;  &quot;VUILD59&quot; 
[31] &quot;VUILD78&quot;  &quot;VUILD91&quot;  &quot;VUILD96&quot; </code></pre>
<pre class="r"><code>intersect(gt_data$tagSNP, exp_data$tagSNP)</code></pre>
<pre><code> [1] &quot;rs10027069&quot;  &quot;rs4693825&quot;   &quot;rs4693826&quot;   &quot;rs11056929&quot;  &quot;rs11056933&quot; 
 [6] &quot;rs214908&quot;    &quot;rs75833251&quot;  &quot;rs230535&quot;    &quot;rs1598858&quot;   &quot;rs230507&quot;   
[11] &quot;rs11202915&quot;  &quot;rs139137711&quot; &quot;rs1926194&quot;   &quot;rs9658742&quot;   &quot;rs2488319&quot;  
[16] &quot;rs2245844&quot;   &quot;rs57721459&quot;  &quot;rs1187102&quot;   &quot;rs35726481&quot;  &quot;rs7340881&quot;  
[21] &quot;rs2518058&quot;   &quot;rs3903056&quot;   &quot;rs7575823&quot;   &quot;rs4853532&quot;   &quot;rs1214596&quot;  
[26] &quot;rs1723018&quot;   &quot;rs1187104&quot;   &quot;rs1187101&quot;   &quot;rs10750646&quot;  &quot;rs12575975&quot; 
[31] &quot;rs11645013&quot;  &quot;rs12445247&quot;  &quot;rs1017734&quot;   &quot;rs1017736&quot;   &quot;rs1187068&quot;  
[36] &quot;rs10085877&quot;  &quot;rs12704057&quot;  &quot;rs17195631&quot;  &quot;rs75045913&quot;  &quot;rs2928672&quot;  
[41] &quot;rs5743462&quot;   &quot;rs2977773&quot;   &quot;rs2293960&quot;   &quot;rs7072966&quot;   &quot;rs2277235&quot;  
[46] &quot;rs3013244&quot;   &quot;rs2619344&quot;   &quot;rs2737017&quot;   &quot;rs56932055&quot;  &quot;rs10484326&quot; 
[51] &quot;rs11252856&quot;  &quot;rs498327&quot;    &quot;rs8056195&quot;   &quot;rs11864751&quot;  &quot;rs3088162&quot;  
[56] &quot;rs71334202&quot;  &quot;rs1029463&quot;   &quot;rs2696609&quot;   &quot;rs371996525&quot; &quot;rs577973995&quot;
[61] &quot;rs558642106&quot; &quot;rs542050069&quot; &quot;rs546433642&quot; &quot;rs563326240&quot; &quot;rs569247938&quot;
[66] &quot;rs549599956&quot; &quot;rs2668670&quot;   &quot;rs4419368&quot;   &quot;rs34705070&quot;  &quot;rs6772168&quot;  
[71] &quot;rs13163344&quot;  &quot;rs138215727&quot; &quot;rs7844887&quot;   &quot;rs35318617&quot;  &quot;rs7083234&quot;  
[76] &quot;rs3890643&quot;   &quot;rs75391810&quot;  &quot;rs718874&quot;    &quot;rs718875&quot;    &quot;rs58888255&quot; 
[81] &quot;rs1608&quot;      &quot;rs7410765&quot;   &quot;rs12171243&quot;  &quot;rs55931899&quot;  &quot;rs68057779&quot; 
[86] &quot;rs12764498&quot;  &quot;rs7608471&quot;   &quot;rs4399765&quot;   &quot;rs75101148&quot;  &quot;rs58606426&quot; 
[91] &quot;rs3934643&quot;   &quot;rs12528271&quot;  &quot;rs2240803&quot;   &quot;rs9600498&quot;   &quot;rs8192757&quot;  </code></pre>
<pre class="r"><code>setdiff(gt_data$tagSNP, exp_data$tagSNP)</code></pre>
<pre><code>[1] &quot;rs28670302&quot; &quot;rs2978483&quot; </code></pre>
<pre class="r"><code>setdiff(exp_data$tagSNP, gt_data$tagSNP)</code></pre>
<pre><code>character(0)</code></pre>
<pre class="r"><code>setdiff(gt_data$Sample_Name, exp_data$Sample_Name)</code></pre>
<pre><code> [1] &quot;THD0001&quot;  &quot;THD0002&quot;  &quot;THD0005&quot;  &quot;THD0006&quot;  &quot;THD0007&quot;  &quot;THD0008&quot; 
 [7] &quot;THD0009&quot;  &quot;THD0010&quot;  &quot;THD0012&quot;  &quot;THD0014&quot;  &quot;THD0015&quot;  &quot;THD0017&quot; 
[13] &quot;THD0019&quot;  &quot;THD0020&quot;  &quot;THD0021&quot;  &quot;THD0022&quot;  &quot;THD0023&quot;  &quot;TILD015&quot; 
[19] &quot;TILD019&quot;  &quot;TILD039&quot;  &quot;TILD051&quot;  &quot;TILD060&quot;  &quot;TILD084&quot;  &quot;TILD116&quot; 
[25] &quot;VUHD071&quot;  &quot;VUHD072&quot;  &quot;VUHD073&quot;  &quot;VUHD077&quot;  &quot;VUHD078&quot;  &quot;VUHD087&quot; 
[31] &quot;VUHD101&quot;  &quot;VUHD103&quot;  &quot;VUHD104&quot;  &quot;VUHD106&quot;  &quot;VUHD107&quot;  &quot;VUHD108&quot; 
[37] &quot;VUHD111&quot;  &quot;VUHD112&quot;  &quot;VUHD114&quot;  &quot;VUHD117&quot;  &quot;VUHD65&quot;   &quot;VUHD66&quot;  
[43] &quot;VUHD67&quot;   &quot;VUHD68&quot;   &quot;VUHD69&quot;   &quot;VUHD70&quot;   &quot;VUHD84&quot;   &quot;VUHD85&quot;  
[49] &quot;VUHD92&quot;   &quot;VUHD94&quot;   &quot;VUHD95&quot;   &quot;VUHD98&quot;   &quot;VUILD100&quot; &quot;VUILD53&quot; 
[55] &quot;VUILD54&quot;  &quot;VUILD55&quot;  &quot;VUILD57&quot;  &quot;VUILD58&quot;  &quot;VUILD60&quot;  &quot;VUILD61&quot; 
[61] &quot;VUILD62&quot;  &quot;VUILD63&quot;  &quot;VUILD64&quot;  &quot;VUILD65&quot;  &quot;VUILD67&quot;  &quot;VUILD68&quot; 
[67] &quot;VUILD69&quot;  &quot;VUILD73&quot;  &quot;VUILD74&quot;  &quot;VUILD75&quot;  &quot;VUILD76&quot;  &quot;VUILD77&quot; 
[73] &quot;VUILD79&quot;  &quot;VUILD80&quot;  &quot;VUILD81&quot;  &quot;VUILD82&quot;  &quot;VUILD83&quot;  &quot;VUILD84&quot; 
[79] &quot;VUILD85&quot;  &quot;VUILD86&quot;  &quot;VUILD87&quot;  &quot;VUILD88&quot;  &quot;VUILD89&quot;  &quot;VUILD90&quot; 
[85] &quot;VUILD92&quot;  &quot;VUILD93&quot;  &quot;VUILD95&quot;  &quot;VUILD97&quot;  &quot;VUILD98&quot;  &quot;VUILD99&quot; </code></pre>
<pre class="r"><code>setdiff(exp_data$Sample_Name, gt_data$Sample_Name)</code></pre>
<pre><code>character(0)</code></pre>
<pre class="r"><code>gt_data &lt;- as.data.frame(gt_data)

exp_gt_data &lt;- exp_data %&gt;% full_join(gt_data, by = c(&quot;tagSNP&quot;, &quot;Sample_Name&quot;))</code></pre>
</div>
<div id="plotting-by-genotype" class="section level3">
<h3>Plotting by genotype</h3>
<pre class="r"><code>summary(exp_gt_data$ALTprobeCounts)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s 
  0.000   0.003   0.007   0.294   0.026   7.243    8796 </code></pre>
<pre class="r"><code>summary(exp_gt_data$REFprobeCounts)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s 
  0.000   0.003   0.007   0.359   0.022  13.588    8796 </code></pre>
<pre class="r"><code>exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  ggplot(aes(y = ALTprobeCounts, x = REFprobeCounts, color = eSNP_dosage)) + 
  geom_point() +
  theme_classic()</code></pre>
<pre><code>Warning: Removed 8796 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  ggplot(aes(y = log(ALTprobeCounts), x = log(REFprobeCounts), color = eSNP_dosage)) + 
  geom_point() +
  theme_classic()</code></pre>
<pre><code>Warning: Removed 8796 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-8-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_gt_data %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                             names_to = &quot;which_probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;% #dplyr::select(aveCounts) %&gt;% unique()
  ggplot(aes(y = log(aveCounts), x = which_probe, fill = which_probe)) + 
    geom_violin(position = position_dodge(width = 0.9)) +
    #geom_point(position = position_jitterdodge(seed = 1, dodge.width = 0.9)) +
    theme_classic() +
    xlab(&quot;eSNP gt&quot;) +
    ylab(&quot;Average log(CPM)&quot;)</code></pre>
<pre><code>Warning: Removed 17641 rows containing non-finite values (`stat_ydensity()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-8-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_gt_data %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                             names_to = &quot;which_probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;% #dplyr::select(aveCounts) %&gt;% unique()
  ggplot(aes(y = log(aveCounts), x = eSNP_dosage, fill = which_probe)) + 
    geom_violin(position = position_dodge(width = 0.9)) +
    #geom_point(position = position_jitterdodge(seed = 1, dodge.width = 0.9)) +
    theme_classic() +
    xlab(&quot;eSNP gt&quot;) +
    ylab(&quot;Average log(CPM)&quot;)</code></pre>
<pre><code>Warning: Removed 17641 rows containing non-finite values (`stat_ydensity()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-8-4.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="examining-instances-where-alt-probe-reads-ref-probe-reads"
class="section level3">
<h3>Examining instances where ALT probe reads &gt; REF probe reads</h3>
<pre class="r"><code>exp_gt_data %&gt;% filter(ALTprobeCounts &gt; REFprobeCounts) %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                             names_to = &quot;which_probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;% #dplyr::select(aveCounts) %&gt;% unique()
  ggplot(aes(y = log(aveCounts), x = eSNP_dosage, fill = which_probe)) + 
    geom_violin(position = position_dodge(width = 0.9)) +
    #geom_point(position = position_jitterdodge(seed = 1, dodge.width = 0.9)) +
    theme_classic() +
    xlab(&quot;eSNP gt&quot;) +
    ylab(&quot;Average log(counts)&quot;) +
    ggtitle(&quot;Samples/loci where ALTprobeCounts &gt; REFprobeCounts&quot;)</code></pre>
<pre><code>Warning: Removed 27 rows containing non-finite values (`stat_ydensity()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_gt_data %&gt;% filter(REFprobeCounts &gt; ALTprobeCounts) %&gt;% pivot_longer(cols = c(&quot;ALTprobeCounts&quot;, &quot;REFprobeCounts&quot;),
                             names_to = &quot;which_probe&quot;, values_to = &quot;aveCounts&quot;) %&gt;%
  mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;% #dplyr::select(aveCounts) %&gt;% unique()
  ggplot(aes(y = log(aveCounts), x = eSNP_dosage, fill = which_probe)) + 
    geom_violin(position = position_dodge(width = 0.9)) +
    #geom_point(position = position_jitterdodge(seed = 1, dodge.width = 0.9)) +
    theme_classic() +
    xlab(&quot;eSNP gt&quot;) +
    ylab(&quot;Average log(counts)&quot;) +
    ggtitle(&quot;Samples/loci where REFprobeCounts &gt; ALTprobeCounts&quot;)</code></pre>
<pre><code>Warning: Removed 20 rows containing non-finite values (`stat_ydensity()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-9-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="inspecting-subsets-from-the-scatterplot"
class="section level3">
<h3>Inspecting subsets from the scatterplot</h3>
<pre class="r"><code>exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  ggplot(aes(y = ALTprobeCounts, x = REFprobeCounts, color = eSNP_dosage)) + 
  geom_point() +
  theme_classic()</code></pre>
<pre><code>Warning: Removed 8796 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Probes with samples with high ALT counts and zero REF counts
exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  ggplot(aes(y = ALTprobeCounts, x = REFprobeCounts, color = eSNP_dosage)) + 
  geom_point() +
  theme_classic() +
  geom_vline(xintercept = 0.2) +
  geom_hline(yintercept = 1.8)</code></pre>
<pre><code>Warning: Removed 8796 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  filter(REFprobeCounts &lt; 0.2,
         ALTprobeCounts &gt; 1.8) %&gt;%
  dplyr::select(c(&quot;gene.x&quot;, &quot;tagSNP&quot;, &quot;eSNP&quot;, &quot;Sample_Name&quot;, &quot;probe&quot;)) %&gt;%
  distinct()</code></pre>
<pre><code>    gene.x     tagSNP      eSNP Sample_Name                probe
1  SPARCL1 rs10027069 rs1049544     TILD001 SPARCL1-4-87503683-T
2    STAT1  rs4853532 rs1400657     THD0016  STAT1-2-190971945-T
3    STAT1  rs4853532 rs1400657     THD0026  STAT1-2-190971945-T
4    STAT1  rs4853532 rs1400657     TILD001  STAT1-2-190971945-T
5    STAT1  rs4853532 rs1400657     TILD006  STAT1-2-190971945-T
6    STAT1  rs4853532 rs1400657     TILD010  STAT1-2-190971945-T
7    STAT1  rs4853532 rs1400657     TILD028  STAT1-2-190971945-T
8    STAT1  rs4853532 rs1400657     TILD030  STAT1-2-190971945-T
9    STAT1  rs4853532 rs1400657     TILD037  STAT1-2-190971945-T
10   STAT1  rs4853532 rs1400657     TILD041  STAT1-2-190971945-T
11   STAT1  rs4853532 rs1400657     TILD049  STAT1-2-190971945-T
12   STAT1  rs4853532 rs1400657     TILD055  STAT1-2-190971945-T
13   STAT1  rs4853532 rs1400657     TILD059  STAT1-2-190971945-T
14   STAT1  rs4853532 rs1400657     TILD062  STAT1-2-190971945-T
15   STAT1  rs4853532 rs1400657     TILD074  STAT1-2-190971945-T
16   STAT1  rs4853532 rs1400657     TILD080  STAT1-2-190971945-T
17   STAT1  rs4853532 rs1400657     TILD093  STAT1-2-190971945-T
18   STAT1  rs4853532 rs1400657     TILD103  STAT1-2-190971945-T
19   STAT1  rs4853532 rs1400657     TILD109  STAT1-2-190971945-T
20   STAT1  rs4853532 rs1400657     TILD111  STAT1-2-190971945-T
21   STAT1  rs4853532 rs1400657     TILD113  STAT1-2-190971945-T
22   STAT1  rs4853532 rs1400657     TILD123  STAT1-2-190971945-T
23   STAT1  rs4853532 rs1400657     TILD126  STAT1-2-190971945-T
24   STAT1  rs4853532 rs1400657     TILD136  STAT1-2-190971945-T
25   STAT1  rs4853532 rs1400657     VUHD113  STAT1-2-190971945-T
26   STAT1  rs4853532 rs1400657     VUHD115  STAT1-2-190971945-T
27   STAT1  rs4853532 rs1400657    VUILD102  STAT1-2-190971945-T
28   STAT1  rs4853532 rs1400657    VUILD104  STAT1-2-190971945-T
29   STAT1  rs4853532 rs1400657     VUILD48  STAT1-2-190971945-T
30   STAT1  rs4853532 rs1400657     VUILD59  STAT1-2-190971945-T
31   STAT1  rs4853532 rs1400657     VUILD78  STAT1-2-190971945-T
32   STAT1  rs4853532 rs1400657     VUILD91  STAT1-2-190971945-T
33   STAT1  rs4853532 rs1400657     VUILD96  STAT1-2-190971945-T</code></pre>
<pre class="r"><code># Plotting STAT1
exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  filter(gene.x == &quot;STAT1&quot;,
         eSNP == &quot;rs1400657&quot;) %&gt;%
  ggplot(aes(x = eSNP_dosage, y = ALTprobeCounts/REFprobeCounts, color = eSNP_dosage)) +
  geom_violin() +
  theme_classic()</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-10-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>exp_gt_data %&gt;% mutate(eSNP_dosage = as.character(eSNP_GT)) %&gt;%
  filter(gene.x == &quot;STAT1&quot;,
         eSNP == &quot;rs1400657&quot;) %&gt;%
  ggplot(aes(y = ALTprobeCounts, x = REFprobeCounts, color = eSNP_dosage)) + 
  geom_point() +
  theme_classic()</code></pre>
<p><img src="figure/aggregate_multiSNP.Rmd/unnamed-chunk-10-4.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="todo-aggregating-refalt-for-eqtl-with-multiple-tag-snps"
class="section level3">
<h3>TODO: Aggregating REF/ALT for eQTL with multiple tag-SNPs</h3>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ComplexHeatmap_2.18.0       RColorBrewer_1.1-3         
 [3] viridis_0.6.3               viridisLite_0.4.2          
 [5] ggthemes_4.2.4              VariantAnnotation_1.48.1   
 [7] Rsamtools_2.18.0            Biostrings_2.70.2          
 [9] XVector_0.42.0              SummarizedExperiment_1.32.0
[11] Biobase_2.62.0              GenomicRanges_1.54.1       
[13] GenomeInfoDb_1.38.7         IRanges_2.36.0             
[15] S4Vectors_0.40.2            MatrixGenerics_1.14.0      
[17] matrixStats_1.0.0           BiocGenerics_0.48.1        
[19] googlesheets4_1.1.0         workflowr_1.7.1            
[21] ggrepel_0.9.3               ggpubr_0.6.0               
[23] plyr_1.8.8                  lubridate_1.9.2            
[25] forcats_1.0.0               stringr_1.5.0              
[27] dplyr_1.1.2                 purrr_1.0.2                
[29] readr_2.1.4                 tidyr_1.3.0                
[31] tibble_3.2.1                ggplot2_3.4.2              
[33] tidyverse_2.0.0             SeuratDisk_0.0.0.9021      
[35] Seurat_5.0.1                SeuratObject_5.0.1         
[37] sp_1.6-1                    cli_3.6.1                  

loaded via a namespace (and not attached):
  [1] fs_1.6.2                 spatstat.sparse_3.0-1    bitops_1.0-7            
  [4] doParallel_1.0.17        httr_1.4.6               tools_4.3.0             
  [7] sctransform_0.4.1        backports_1.4.1          utf8_1.2.3              
 [10] R6_2.5.1                 lazyeval_0.2.2           uwot_0.1.14             
 [13] GetoptLong_1.0.5         withr_2.5.0              prettyunits_1.1.1       
 [16] gridExtra_2.3            progressr_0.13.0         spatstat.explore_3.2-1  
 [19] fastDummies_1.7.3        labeling_0.4.2           sass_0.4.6              
 [22] spatstat.data_3.0-1      ggridges_0.5.4           pbapply_1.7-0           
 [25] parallelly_1.36.0        BSgenome_1.70.1          rstudioapi_0.14         
 [28] RSQLite_2.3.1            shape_1.4.6              generics_0.1.3          
 [31] BiocIO_1.12.0            ica_1.0-3                spatstat.random_3.1-5   
 [34] car_3.1-2                Matrix_1.6-5             fansi_1.0.4             
 [37] abind_1.4-5              lifecycle_1.0.3          whisker_0.4.1           
 [40] yaml_2.3.7               carData_3.0-5            SparseArray_1.2.4       
 [43] BiocFileCache_2.10.1     Rtsne_0.16               blob_1.2.4              
 [46] promises_1.2.0.1         crayon_1.5.2             miniUI_0.1.1.1          
 [49] lattice_0.21-8           cowplot_1.1.1            GenomicFeatures_1.54.3  
 [52] KEGGREST_1.42.0          pillar_1.9.0             knitr_1.43              
 [55] rjson_0.2.21             future.apply_1.11.0      codetools_0.2-19        
 [58] leiden_0.4.3             glue_1.6.2               getPass_0.2-4           
 [61] data.table_1.14.8        vctrs_0.6.5              png_0.1-8               
 [64] spam_2.9-1               cellranger_1.1.0         gtable_0.3.3            
 [67] cachem_1.0.8             xfun_0.39                S4Arrays_1.2.1          
 [70] mime_0.12                survival_3.5-5           gargle_1.4.0            
 [73] iterators_1.0.14         ellipsis_0.3.2           fitdistrplus_1.1-11     
 [76] ROCR_1.0-11              nlme_3.1-162             bit64_4.0.5             
 [79] progress_1.2.2           filelock_1.0.2           RcppAnnoy_0.0.20        
 [82] rprojroot_2.0.3          bslib_0.4.2              irlba_2.3.5.1           
 [85] KernSmooth_2.23-21       colorspace_2.1-0         DBI_1.1.3               
 [88] tidyselect_1.2.0         processx_3.8.1           bit_4.0.5               
 [91] compiler_4.3.0           curl_5.0.1               git2r_0.32.0            
 [94] hdf5r_1.3.8              xml2_1.3.4               DelayedArray_0.28.0     
 [97] plotly_4.10.2            rtracklayer_1.62.0       scales_1.2.1            
[100] lmtest_0.9-40            callr_3.7.3              rappdirs_0.3.3          
[103] digest_0.6.31            goftest_1.2-3            spatstat.utils_3.0-3    
[106] rmarkdown_2.22           htmltools_0.5.5          pkgconfig_2.0.3         
[109] highr_0.10               dbplyr_2.3.2             fastmap_1.1.1           
[112] GlobalOptions_0.1.2      rlang_1.1.1              htmlwidgets_1.6.2       
[115] shiny_1.7.4              farver_2.1.1             jquerylib_0.1.4         
[118] zoo_1.8-12               jsonlite_1.8.5           BiocParallel_1.36.0     
[121] RCurl_1.98-1.12          magrittr_2.0.3           GenomeInfoDbData_1.2.11 
[124] dotCall64_1.0-2          patchwork_1.1.2          munsell_0.5.0           
[127] Rcpp_1.0.10              reticulate_1.29          stringi_1.7.12          
[130] zlibbioc_1.48.0          MASS_7.3-60              parallel_4.3.0          
[133] listenv_0.9.0            deldir_1.0-9             splines_4.3.0           
[136] tensor_1.5               circlize_0.4.15          hms_1.1.3               
[139] ps_1.7.5                 igraph_1.4.3             spatstat.geom_3.2-1     
[142] ggsignif_0.6.4           RcppHNSW_0.5.0           reshape2_1.4.4          
[145] biomaRt_2.58.2           XML_3.99-0.14            evaluate_0.21           
[148] foreach_1.5.2            tzdb_0.4.0               httpuv_1.6.11           
[151] RANN_2.6.1               polyclip_1.10-4          clue_0.3-64             
[154] future_1.32.0            scattermore_1.2          broom_1.0.4             
[157] xtable_1.8-4             restfulr_0.0.15          RSpectra_0.16-1         
[160] rstatix_0.7.2            later_1.3.1              googledrive_2.1.0       
[163] GenomicAlignments_1.38.2 memoise_2.0.1            AnnotationDbi_1.64.1    
[166] cluster_2.1.4            timechange_0.2.0         globals_0.16.2          </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
