---
title: "Sample-level pseudobulk and ASE"
author: "heinin"
date: "2024-01-08"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

### Packages and environment variables

```{r, results='hide', warning=F, message=F}

suppressPackageStartupMessages({library(cli)
                                library(Seurat)
                                library(SeuratObject)
                                library(SeuratDisk)
                                library(tidyverse)
                                library(tibble)
                                library(plyr)
                                library(dplyr)
                                library(ggplot2)
                                library(ggpubr)
                                library(ggrepel)
                                library(workflowr)
                                library(googlesheets4)
                                library(VariantAnnotation)})

setwd("/home/hnatri/ILD_ASE_Xenium/")
set.seed(9999)
options(ggrepel.max.overlaps = Inf)

# Colors, themes, cell type markers, and plot functions
source("/home/hnatri/ILD_ASE_Xenium/code/colors_themes.R")
source("/home/hnatri/ILD_ASE_Xenium/code/plot_functions.R")
source("/home/hnatri/ILD_ASE_Xenium/code/utilities.R")

# Cell type annotations
gs4_deauth()
ct_annot  <- gs4_get("https://docs.google.com/spreadsheets/d/1SDfhxf6SjllxXEtNPf32ZKTEqHC9QJW3BpRYZFhpqFE/edit?usp=sharing")
sheet_names(ct_annot)
all_celltypes_annot <- read_sheet(ct_annot, sheet = "All celltypes, annotated, merged")

```

### Importing data

```{r}

seurat_object <- readRDS("/tgen_labs/banovich/IPF/Spatial_ASE/ILD_ASE_Xenium_annotated.rds")

# Genotype data from ASE.rds
gt_data <- read.csv("/scratch/hnatri/ILD/ILD_spatial_ASE/gt_data.csv", row.names = "X")

```

### Importing eQTL and tag-SNP info

```{r}

gs4_deauth()
gtables  <- gs4_get("https://docs.google.com/spreadsheets/d/1f7HtQC07PWa1ij9UxTmgrEMX2OnHDVcrpn7Oa_VNIms/edit?usp=sharing")
sheet_names(gtables)
ld_snp_info <- read_sheet(gtables, sheet = "More candidates, SNP info with exon number")
ld_snp_info <- as.data.frame(ld_snp_info)

# Adding probe names
ld_snp_info$probe <- paste(ld_snp_info$gene, ld_snp_info$chr, ld_snp_info$linked_snp_pos, ld_snp_info$alt, sep = "-")
ld_snp_info$probe1 <- paste(ld_snp_info$gene, ld_snp_info$chr, ld_snp_info$linked_snp_pos, ld_snp_info$ref, sep = "-")
ld_snp_info$probe2 <- paste(ld_snp_info$gene, ld_snp_info$chr, ld_snp_info$linked_snp_pos, ld_snp_info$alt, sep = "-")

```

### Obtaining pseudo-bulk expression values

```{r}

pseudobulk <- AverageExpression(seurat_object,
                                group.by = "Sample_Name",
                                assay = "RNA",
                                slot = "counts")

pseudobulk <- as.data.frame(pseudobulk$RNA)

```

### Looping through each gene-tag-SNP -pair and finding the allelic ratio

ASE is shown as the ALT/REF ratio of read counts.

```{r}

ld_snp_info$gene_tagSNP <- paste0(ld_snp_info$gene, "_", ld_snp_info$gene_body_snp)
gene_tagSNP <- unique(ld_snp_info$gene_tagSNP)

# Creating a dataframe for the results
ratio_res <- data.frame(matrix(NA, ncol = ncol(pseudobulk),
                              nrow = length(gene_tagSNP)))
rownames(ratio_res) <- gene_tagSNP
colnames(ratio_res) <- colnames(pseudobulk)

# Looping through every locus, calculating the allelic ratio
for(i in rownames(ratio_res)){
  # Finding the probe names for REF and ALT
  locus_info <- ld_snp_info %>%
    filter(gene_tagSNP == i) %>%
    dplyr::select(-c("esnp", "gene_snp", "R2", "esnp_loc", "type")) %>%
    distinct()
  # Calculating the ratio
  ref_exp <- pseudobulk[locus_info$probe1,]
  alt_exp <- pseudobulk[locus_info$probe2,]
  ratio_res[i,] <- alt_exp/ref_exp
}

saveRDS(ratio_res, "/scratch/hnatri/ILD/ILD_spatial_ASE/sample_ase_res.rds")

```

### Applyin a Wilcoxon rank-sum test to calculate the statistical significance 
of the differential ASE between genotypes

```{r, results='hide', warning=F, message=F}

# Creating a dataframe for the results
wilcox_res <- data.frame(matrix(NA, ncol = 4, nrow = length(gene_tagSNP)))
rownames(wilcox_res) <- gene_tagSNP
colnames(wilcox_res) <- c("p_2vs0", "p_1_2vs0", "p_adj_2vs0", "p_adj_1_2vs0")

# Looping through every locus, calculating the allelic ratio
for(i in rownames(wilcox_res)){
  tagSNP <- strsplit(i, split = "_")[[1]][2]
  
  # ASE values
  ase_vals <- ratio_res %>% filter(row.names(ratio_res) == i) %>%
    t() %>% as.data.frame() %>%
    rownames_to_column(var = "Sample_Name")
  ase_vals <- ase_vals[complete.cases(ase_vals),]
  ase_vals <- ase_vals %>%
    filter_if(~is.numeric(.), all_vars(!is.infinite(.)))
  # Adding genotype information
  gt_vals <- gt_data %>% filter(gene_body_snp == tagSNP)
  # gt_vals[which(gt_vals$linked_SNP_GT==0), "Sample_Name"]
  ase_00 <- ase_vals %>% filter(Sample_Name %in% gt_vals[which(gt_vals$linked_SNP_GT==0), "Sample_Name"])
  ase_01 <- ase_vals %>% filter(Sample_Name %in% gt_vals[which(gt_vals$linked_SNP_GT==1), "Sample_Name"])
  ase_11 <- ase_vals %>% filter(Sample_Name %in% gt_vals[which(gt_vals$linked_SNP_GT==2), "Sample_Name"])
  
  if(length(unique(c(ase_00[,i], ase_01[,i], ase_11[,i])))<=1){
    next
  } else if (nrow(ase_11)==0 & nrow(ase_01)==0){
    next
  } else if (nrow(ase_00)==0){
    next
  } else if (nrow(ase_11)==0 | nrow(ase_01)==0){
    wres1 <- wilcox.test(c(ase_11[,i], ase_01[,i]), ase_00[,i], alternative = "two.sided")
    wilcox_res[i, c("p_1_2vs0")] <- wres1[["p.value"]]
    
    next
  } else {
    wres1 <- wilcox.test(c(ase_11[,i], ase_01[,i]), ase_00[,i], alternative = "two.sided")
    wres2 <- wilcox.test(ase_11[,i], ase_00[,i], alternative = "two.sided")
  
    wilcox_res[i, c("p_1_2vs0")] <- wres1[["p.value"]]
    wilcox_res[i, c("p_2vs0")] <- wres2[["p.value"]]
  }
}

# Adjusting for multiple comparisons
wilcox_res["p_adj_1_2vs0"] <- p.adjust(wilcox_res[,"p_1_2vs0"],
                                       n = length(wilcox_res[,"p_1_2vs0"]),
                                       method = "fdr")
wilcox_res["p_adj_2vs0"] <- p.adjust(wilcox_res[,"p_2vs0"],
                                     n = length(wilcox_res[,"p_2vs0"]),
                                     method = "fdr")

# Saving the results for later
saveRDS(wilcox_res, "/scratch/hnatri/ILD/ILD_spatial_ASE/sample_ase_sigtest.rds")

```

### Plotting the *p*-value distribution

```{r, fig.height=4, results='hide', warning=F, message=F}

# Nominal p
hist_1_2vs0 <- ggplot(wilcox_res, aes(x = p_1_2vs0)) +
                      geom_histogram(binwidth = 0.1) +
                      xlim(0, 1.05) +
                      xlab("0/1 & 1/1 vs. 0/0, nominal p") +
                      theme_bw() +
                      my_theme +
                      NoLegend()

hist_2vs0 <- ggplot(wilcox_res, aes(x = p_2vs0)) +
                    geom_histogram(binwidth = 0.1) +
                    xlim(0, 1.05) +
                    xlab("1/1 vs. 0/0, nominal p") +
                    theme_bw() +
                    my_theme +
                    NoLegend()

hist_1_2vs0 + hist_2vs0

# FDR-p
hist_adj_1_2vs0 <- ggplot(wilcox_res, aes(x = p_adj_1_2vs0)) +
                          geom_histogram(binwidth = 0.1) +
                          xlim(0, 1.05) +
                          xlab("0/1 & 1/1 vs. 0/0, FDR-p") +
                          theme_bw() +
                          my_theme +
                          NoLegend()

hist_adj_2vs0 <- ggplot(wilcox_res, aes(x = p_adj_2vs0)) +
                        geom_histogram(binwidth = 0.1) +
                        xlim(0, 1.05) +
                        xlab("1/1 vs. 0/0, FDR-p") +
                        theme_bw() +
                        my_theme +
                        NoLegend()

hist_adj_1_2vs0 + hist_adj_2vs0

```

### Examining the most significant loci

```{r}

# Subsetting significant loci
wilcox_res %>% filter(p_adj_1_2vs0 <= 0.05 | p_adj_2vs0 <= 0.05)

```

